<div style="background:linear-gradient(135deg,#7a3d00 0%,#b45309 55%,#d97706 100%);border-radius:18px;padding:34px 30px;color:#fff;font-family:Inter,Segoe UI,sans-serif">
  <div style="font-size:12px;letter-spacing:3px;color:#ffe9c7;font-weight:700;text-transform:uppercase">Chapter 7 · Solutions</div>
  <div style="font-size:36px;font-weight:900;line-height:1.1;margin:10px 0 6px">Practice Challenges, Worked Answers ✅</div>
  <div style="font-size:15px;color:#ffe6cc;max-width:700px;line-height:1.6">Full solutions to the five "Math Refresher" challenges, one per math area. Try them yourself first, then compare.</div>
  <div style="margin-top:16px;font-size:13px;color:#ffe2bf">Statistics, Data Science and AI: A Visual Handbook · John Fisher · 2026</div>
</div>

## ⚙️ Setup

In [1]:
import numpy as np
import math
import pandas as pd
print("✅ Ready.")

✅ Ready.


<div style="background:#fef3e2;border-left:5px solid #d97706;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#0891b2;letter-spacing:1px">CHALLENGE 1 · SUMMATION</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Compute Σ by hand and in code</div>
<div style="color:#4a5578;margin-top:6px">For x = [3, 7, 2, 8]: compute Σxᵢ, then the mean (1/n)·Σxᵢ. Verify against NumPy.</div>
</div>

In [2]:
x = np.array([3, 7, 2, 8])
total = x.sum()
mean = total / len(x)
print(f"Σ xᵢ  = {total}")
print(f"mean  = {mean}   |  np.mean = {np.mean(x)}")

Σ xᵢ  = 20
mean  = 5.0   |  np.mean = 5.0


**Answer:** Σxᵢ = **20**, mean = 20/4 = **5.0**. Σ is just "add them all"; dividing by n gives the mean.

<div style="background:#fef3e2;border-left:5px solid #d97706;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#7c3aed;letter-spacing:1px">CHALLENGE 2 · SETS</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Union and intersection</div>
<div style="color:#4a5578;margin-top:6px">Let A = {1,2,3,4} and B = {3,4,5,6}. Find A∪B (or), A∩B (and), and how many elements are in the union.</div>
</div>

In [3]:
A = {1, 2, 3, 4}; B = {3, 4, 5, 6}
print("A ∪ B =", A | B)
print("A ∩ B =", A & B)
print("size of union =", len(A | B))

A ∪ B = {1, 2, 3, 4, 5, 6}
A ∩ B = {3, 4}
size of union = 6


**Answer:** A∪B = {1,2,3,4,5,6} (6 elements), A∩B = {3,4}. In probability, ∪ is "or" and ∩ is "and".

<div style="background:#fef3e2;border-left:5px solid #d97706;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#d97706;letter-spacing:1px">CHALLENGE 3 · COUNTING</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Permutation or combination?</div>
<div style="color:#4a5578;margin-top:6px">(a) How many ways can 4 runners finish 1st/2nd/3rd? (b) How many 3-person teams can you form from 4 people?</div>
</div>

In [4]:
# (a) order matters (podium positions) -> permutation P(4,3)
print("(a) finishing orders  P(4,3) =", math.perm(4,3))
# (b) order does NOT matter (a team) -> combination C(4,3)
print("(b) possible teams    C(4,3) =", math.comb(4,3))

(a) finishing orders  P(4,3) = 24
(b) possible teams    C(4,3) = 4


**Answer:** (a) **24** ordered finishes (order matters → permutation). (b) **4** teams (order doesn't matter → combination).

<div style="background:#fef3e2;border-left:5px solid #d97706;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#059669;letter-spacing:1px">CHALLENGE 4 · LOGARITHMS</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Products into sums</div>
<div style="color:#4a5578;margin-top:6px">Show that log(8 × 32) equals log(8) + log(32). Then: what is log₂(1024)?</div>
</div>

In [5]:
print("log(8·32)        =", round(np.log(8*32), 4))
print("log(8) + log(32) =", round(np.log(8) + np.log(32), 4))

print("\nlog₂(1024) =", int(np.log2(1024)), "  (because 2¹⁰ = 1024)")

log(8·32)        = 5.5452
log(8) + log(32) = 5.5452

log₂(1024) = 10   (because 2¹⁰ = 1024)


**Answer:** Both equal ≈ 5.545, logs turn multiplication into addition. And log₂(1024) = **10**, since 2¹⁰ = 1024.

<div style="background:#fef3e2;border-left:5px solid #d97706;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#2563eb;letter-spacing:1px">CHALLENGE 5 · SLOPES & MATRICES</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Derivative and a matrix mean</div>
<div style="color:#4a5578;margin-top:6px">(a) The loss is f(x) = (x−5)². Its derivative is f'(x) = 2(x−5). What is the slope at x=2, and which way is downhill? (b) For the matrix X below, compute the mean of each column.</div>
</div>

In [6]:
# (a) derivative / slope
slope_at_2 = 2*(2 - 5)
print(f"(a) f'(2) = {slope_at_2}  -> slope is negative, so downhill is to the RIGHT (toward x=5).")

# (b) column means of a data matrix
X = np.array([[2, 10],
              [4, 20],
              [6, 30]])
print("(b) column means:", X.mean(axis=0))

(a) f'(2) = -6  -> slope is negative, so downhill is to the RIGHT (toward x=5).
(b) column means: [ 4. 20.]


**Answer:** (a) f'(2) = **−6**; a negative slope means the function is falling, so the minimum (x=5) is to the right, gradient descent would step that way. (b) Column means = **[4, 20]**. Linear algebra averages the whole column at once.

---
<div style="background:#ffffff;border:1px solid #e6e9f2;border-radius:16px;padding:22px 26px;font-family:Inter,sans-serif">
<div style="font-size:19px;font-weight:800;color:#1a2138">🎉 Nicely done!</div>
<div style="color:#4a5578;line-height:1.8;margin-top:8px">You worked through summation, sets, counting, logarithms, slopes, and matrices, the exact math the rest of the book leans on. With Part I complete, you're ready to start describing data.</div>
</div>

<div style="text-align:center;color:#8b94b3;font-size:12px;margin-top:16px">Statistics, Data Science and AI: A Visual Handbook · © 2026 John Fisher</div>